api/repository.go

```diff
@@ -5,6 +5,71 @@ import (
 	"encoding/json"
 )
 
+// RepositoryRaw is the store model for a Repository.
+// Fields have exactly the same meanings as Repository.
+type RepositoryRaw struct {
+	ID int
+
+	// Standard fields
+	CreatorID int
+	CreatedTs int64
+	UpdaterID int
+	UpdatedTs int64
+
+	// Related fields
+	VCSID     int
+	ProjectID int
+
+	// Domain specific fields
+	Name               string
+	FullPath           string
+	WebURL             string
+	BranchFilter       string
+	BaseDirectory      string
+	FilePathTemplate   string
+	SchemaPathTemplate string
+	ExternalID         string
+	ExternalWebhookID  string
+	WebhookURLHost     string
+	WebhookEndpointID  string
+	WebhookSecretToken string
+	AccessToken        string
+	ExpiresTs          int64
+	RefreshToken       string
+}
+
+// ToRepository creates an instance of Repository based on the RepositoryRaw.
+// This is intended to be called when we need to compose a Repository relationship.
+func (raw *RepositoryRaw) ToRepository() *Repository {
```

Comment:
I am thinking of moving the compose logic inside ToRepository, so the signature is ToRepository(s server)

This prevents someone only calling ToRepository and forgetting calling composeXXX, which defeats the purpose of having separate types.

Put in another way, The only way to construct Repository is via ToRepository and it guarantees it's been composed upon returning if no error occurs.

wdyt?

Reply:
@d-bytebase and I had a conversation about a similar problem here: https://github.com/bytebase/bytebase/pull/754#discussion_r820572223

IMHO, this is another big refactor which probably needs some detailed design in the front. I suggest leave later, which is a `TODO` item pointing to the link above here: https://github.com/bytebase/bytebase/issues/758

> This prevents someone only calling ToRepository and forgetting calling composeXXX, which defeats the purpose of having separate types.

The problem exists indeed. But we could get all the good stuff back as long as we figure out how to ensure all the server data models are guaranteed to be composed at the consumer side, and do not import unnecessary composition of unused fields which wastes CPU and memory resources (especially for big struct with multiple relationship pointers).

Reply:
I've updated this comments to #758 as well.

Reply:
OK, BTW, I am not worried about wasted CPU and memory resources at this point.

Whatever refactor we choose, the No 1 priority is one can tell whether the object is composed or not by just looking at the type.

Reply:
+1 we're a management service that doesn't need to worry about resource usage and efficiency as long as it's not causing seconds of delay.